src/dns/local.rs

```diff
@@ -1,134 +1,166 @@
-use std::collections::HashMap;
 use std::net::IpAddr;
 use wildmatch::WildMatch;
-use crate::dns::DNSEntry;
+use crate::dns::{DNSEntry, ResolveType};
+use crate::types;
 
 /// Local override table that can be used instead of using /etc/hosts or similar 3rd party dns system.
+#[derive(Default)]
 pub struct LocalOverrideTable {
     /// Entries in the local override table. First iteration of the resolver: a simple list that
     /// will be queried O(n). Later iterations will use a more efficient data structure like a tree
     /// per domain part (e.g. com, org, net, etc.)
     entries: Vec<DNSEntry>,
-    round_robin_ptrs: HashMap<String, usize>,
 }
 
-impl Default for LocalOverrideTable {
-    fn default() -> LocalOverrideTable {
-        LocalOverrideTable {
+impl LocalOverrideTable {
+
+    pub fn new() -> LocalOverrideTable {
+        let mut table = LocalOverrideTable {
             entries: Vec::new(),
-            round_robin_ptrs: HashMap::new(),
-        }
+        };
+
+        table.reload_table_entries();
+        table
     }
-}
 
-impl LocalOverrideTable {
+    /// Regenerates the new entries table
+    pub fn reload_table_entries(&mut self) {
+        // @todo: this should reload all table entries from the configuration into the self.entries list
+    }
 
-    pub fn resolve(&mut self, domain: &str) -> Option<IpAddr> {
-        let mut ips = Vec::new();
-        let mut wildcard_ips = Vec::new();
+    pub fn resolve(&mut self, domain: &str, resolve_type: ResolveType) -> Result<IpAddr, types::Error> {
+        let mut matched_entry = None;
+        let mut matched_wildcard_entries = Vec::new();
 
         // For now it's ok to iterate the entries list, but later we should use a more efficient
-        // data structure.
-        for entry in &self.entries {
-            // We separate wildcard matches and specific matches.
+        // data structure in the future
+        for (idx, entry) in self.entries.iter().enumerate() {
+            // We separate wildcard matches and specific matches (e.g. *.example.com vs example.com)
             if WildMatch::new(&entry.domain_name).matches(domain) {
-                wildcard_ips.push(entry.ip);
+                matched_wildcard_entries.push(idx);
             }
             if entry.domain_name == domain {
```

Comment:
Maybe put this over the wildcard match because it can optimize the time this a tiny bit. (But feel free to ignore)